packages/shared/ReactTreeTraversal.js

```diff
@@ -5,7 +5,35 @@
  * LICENSE file in the root directory of this source tree.
  */
 
-import {HostComponent} from './ReactWorkTags';
+import {HostComponent, HostPortal, HostRoot} from './ReactWorkTags';
+import {enableModernEventSystem} from './ReactFeatureFlags';
+
+function getParentUntilRootOrPortal(inst) {
+  let node = inst.return;
+
+  while (node !== null) {
+    if (node.tag === HostPortal) {
+      let grandNode = node;
+      const portalNode = node.stateNode.containerInfo;
+      while (grandNode !== null) {
+        // If we find a root that is actually a parent in the DOM tree
+        // then we don't continue with getting the parent, as that root
+        // will have its own event listener.
+        if (
+          grandNode.tag === HostRoot &&
+          grandNode.stateNode.containerInfo.contains(portalNode)
+        ) {
+          return null;
```

Comment:
So there are three scenarios:

- The event target is outside the portal so it only affects the root. In this case there's no problem. In either case.
- The event target is in the portal, and the portal itself is outside the root in the DOM. In this case we want it to still bubble through the React tree here which is what traverseTwoPhase normally does.
- The event target is in the portal, and the portal is itself within the root in the DOM. In this case, unless we do something, we risk double firing the event. That's the problem to solve.

The double firing happens because the event can fire using the listeners attached on the root after the portal fires. It can also fire due to the traversal in traverseTwoPhase while it's firing on the portal.

There are two alternative ways to fix this.

- Prevent the traversal from going through to the root in this scenario. I.e. this is the fix you have in this PR.
- Prevent the root event listener from handling the event if the target is in its own portal.

What if instead we prevent the root's event listeners from firing?

Like what if the loop you have in dispatchEventForPluginEventSystem stopped after it found a portal and it wasn't equal to the rootContainer? Then the root's listeners would not be used when the event happens inside a portal. It's always the portal's listeners that handles all events inside a portal.

Reply:
Thank you for the very detailed feedback. I was able to revise my current implementation to address your concerns. Let me know if I've missed anything. :)

Reply:
Posted a follow up PR to show what I mean: https://github.com/facebook/react/pull/18215/